pkg/apis/agent/v1alpha1/agent_types.go

```diff
@@ -127,6 +132,32 @@ type DeploymentSpec struct {
 	Strategy appsv1.DeploymentStrategy `json:"strategy,omitempty"`
 }
 
+type StatefulSetSpec struct {
+	// +kubebuilder:pruning:PreserveUnknownFields
+	PodTemplate corev1.PodTemplateSpec `json:"podTemplate,omitempty"`
+	Replicas    *int32                 `json:"replicas,omitempty"`
+	ServiceName string                 `json:"serviceName,omitempty"`
+
+	// PodManagementPolicy controls how pods are created during initial scale up,
+	// when replacing pods on nodes, or when scaling down. The default policy is
+	// `Parallel`, where pods are created in parallel to match the desired scale
+	// without waiting, and on scale down will delete all pods at once.
+	// The alternative policy is `OrderedReady`, the default for vanilla kubernetes
+	// StatefulSets, where pods are created in increasing order in increasing order
+	// (pod-0, then pod-1, etc.) and the controller will wait until each pod is ready before
+	// continuing. When scaling down, the pods are removed in the opposite order.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=OrderedReady;Parallel
+	// +kubebuilder:default:=Parallel
+	PodManagementPolicy appsv1.PodManagementPolicyType `json:"podManagementPolicy,omitempty"`
+
+	// VolumeClaimTemplates is a list of persistent volume claims to be used by each Pod.
+	// Every claim in this list must have a matching volumeMount in one of the containers defined in the PodTemplate.
+	// Items defined here take precedence over any default claims added by the operator with the same name.
+	// +kubebuilder:validation:Optional
+	VolumeClaimTemplates []corev1.PersistentVolumeClaim `json:"volumeClaimTemplates,omitempty"`
```

Comment:
I wonder if we want to validate that there is no change applied to volume templates, as it's done for Logstash, something along:

```diff
diff --git a/pkg/apis/agent/v1alpha1/validations.go b/pkg/apis/agent/v1alpha1/validations.go
index 12c59c326..4f5f1365a 100644
--- a/pkg/apis/agent/v1alpha1/validations.go
+++ b/pkg/apis/agent/v1alpha1/validations.go
@@ -9,6 +9,7 @@ import (
        "reflect"
        "strings"
 
+       apiequality "k8s.io/apimachinery/pkg/api/equality"
        "k8s.io/apimachinery/pkg/util/validation/field"
 
        commonv1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/common/v1"
@@ -37,6 +38,7 @@ var (
 
        updateChecks = []func(old, curr *Agent) field.ErrorList{
                checkNoDowngrade,
+               checkPVCChanges,
        }
 )
 
@@ -137,6 +139,23 @@ func checkNoDowngrade(prev, curr *Agent) field.ErrorList {
        return commonv1.CheckNoDowngrade(prev.Spec.Version, curr.Spec.Version)
 }
 
+const (
+       pvcImmutableMsg = "Volume claim templates cannot be modified"
+)
+
+// checkPVCChanges ensures no PVCs are changed, as volume claim templates are immutable in StatefulSets.
+func checkPVCChanges(current, proposed *Agent) field.ErrorList {
+       var errs field.ErrorList
+       if current == nil || proposed == nil || current.Spec.StatefulSet == nil || proposed.Spec.StatefulSet == nil {
+               return errs
+       }
+       if !apiequality.Semantic.DeepEqual(current.Spec.StatefulSet.VolumeClaimTemplates, proposed.Spec.StatefulSet.VolumeClaimTemplates) {
+               errs = append(errs, field.Invalid(field.NewPath("spec").Child("volumeClaimTemplates"), proposed.Spec.StatefulSet.VolumeClaimTemplates, pvcImmutableMsg))
+       }
+
+       return errs
+}
+
 func checkSingleConfigSource(a *Agent) field.ErrorList {
        if a.Spec.Config != nil && a.Spec.ConfigRef != nil {
                msg := "Specify at most one of [`config`, `configRef`], not both"
```

Reply:
oh nice catch, yes I think we want to validate this one. Only one minor change in the proposed commit above, I think we need to check if both current and proposed are both with a `Spec` of `StatefulSet`. Otherwise, we shouldn't check them